File: dotnet/CucumberExpressions.Tests/CucumberExpressionGeneratorTest.cs
using System;
using System.Collections.Generic;
using FluentAssertions;
using Xunit;

namespace CucumberExpressions.Tests;


public class CucumberExpressionGeneratorTest : CucumberExpressionTestBase
{

    private readonly StubParameterTypeRegistry parameterTypeRegistry = new StubParameterTypeRegistry();
    private readonly CucumberExpressionGenerator generator;

    public CucumberExpressionGeneratorTest()
    {
        generator = new CucumberExpressionGenerator(parameterTypeRegistry);
    }

    [Fact]
    public void documents_expression_generation() {
        CucumberExpressionGenerator generator = new CucumberExpressionGenerator(parameterTypeRegistry);
        String undefinedStepText = "I have 2 cucumbers and 1.5 tomato";
        GeneratedExpression generatedExpression = generator.generateExpressions(undefinedStepText)[0];
        Assert.Equal("I have {int} cucumbers and {double} tomato", generatedExpression.getSource());
        Assert.Equal(typeof(double), generatedExpression.getParameterTypes()[1].getType());
    }

    [Fact]
    public void generates_expression_for_no_args() {
        assertExpression("hello", Array.Empty<string>(), "hello");
    }

    [Fact]
    public void generates_expression_with_escaped_left_parenthesis() {
        assertExpression(
                "\\(iii)", Array.Empty<string>(),
                "(iii)");
    }

    [Fact]
    public void generates_expression_with_escaped_left_curly_brace() {
        assertExpression(
                "\\{iii}", Array.Empty<string>(),
                "{iii}");
    }

    [Fact]
    public void generates_expression_with_escaped_slashes() {
        assertExpression(
                "The {int}\\/{int}\\/{int} hey", new[] {"int1", "int2", "int3"},
                "The 1814/05/17 hey");
    }

    [Fact]
    public void generates_expression_for_int_arg() {
        assertExpression(
                "I have {int} cukes", new[] {"int1"},
                "I have 2 cukes");
    }

    [Fact]
    public void generates_expression_for_double_arg() {
        assertExpression(
                "I have {double} cukes", new[] { "double1" },
                "I have 2.5 cukes");
    }

    [Fact]
    public void generates_expression_for_int_double_arg() {
        assertExpression(
                "I have {int} cukes and {double} euro", new[] {"int1", "double1"},
                "I have 2 cukes and 1.5 euro");
    }

    [Fact]
    public void generates_expression_for_numbers_with_symbols_and_currency() {
        assertExpression(
                "Some ${double} of cukes at {int}° Celsius", new[] {"double1", "int1"},
                "Some $5000.00 of cukes at 42° Celsius");
    }

    [Fact]
    public void generates_expression_for_numbers_with_text_on_both_sides() {
        assertExpression(
                "i18n", Array.Empty<string>(),
                "i18n");
    }

    [Fact]
    public void generates_expression_for_strings() {
        assertExpression(
                "I like {string} and {string}", new[] {"string", "string2"},
                "I like \"bangers\" and 'mash'");
    }

    [Fact]
    public void generates_expression_with_percent_sign() {
        assertExpression(
                "I am {int}% foobar", new[] {"int1"},
                "I am 20% foobar");
    }

    [Fact]
    public void generates_expression_for_just_int() {
        assertExpression(
                "{int}", new[] {"int1"},
                "99999");
    }

    [Fact]
    public void numbers_all_arguments_when_type_is_reserved_keyword() {
        assertExpression(
                "I have {int} cukes and {int} euro", new[] {"int1", "int2"},
                "I have 2 cukes and 5 euro");
    }

    [Fact]
    public void numbers_only_second_argument_when_type_is_not_reserved_keyword() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<decimal>(
                "currency",
                "[A-Z]{3}"));
        assertExpression(
                "I have a {currency} account and a {currency} account", new[] {"currency", "currency2"},
                "I have a EUR account and a GBP account");
    }

    [Fact]
    public void does_not_suggest_parameter_type_when_surrounded_by_alphanum() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
                "direction",
                "(up|down)"));
        assertExpression(
                "I like muppets", Array.Empty<string>(),
                "I like muppets");
    }

    [Fact]
    public void does_suggest_parameter_type_when_surrounded_by_space() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "direction",
            "(up|down)"));
        assertExpression(
                "it went {direction} and {direction}", new[] {"direction", "direction2"},
                "it went up and down");
    }

    [Fact]
    public void prefers_leftmost_match_when_there_is_overlap() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "right",
            "c d"));
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "left",
            "b c"));
        assertExpression(
                "a {left} d e f g", new[] {"left"},
                "a b c d e f g");
    }

    [Fact]
    public void prefers_widest_match_when_pos_is_same() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "airport",
            "[A-Z]{3}"));
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "leg",
            "[A-Z]{3}-[A-Z]{3}"));
        assertExpression(
                "leg {leg}", new[] {"leg"},
                "leg LHR-CDG");
    }

    [Fact]
    public void generates_all_combinations_of_expressions_when_several_parameter_types_match() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<decimal>(
            "currency",
            "x"));
        parameterTypeRegistry.defineParameterType(new StubParameterType<DateTime>(
            "date",
            "x"));

        var generatedExpressions = generator.generateExpressions("I have x and x and another x");
        var expressions = new List<String>();
        foreach (GeneratedExpression generatedExpression in generatedExpressions) {
            var source = generatedExpression.getSource();
            expressions.Add(source);
        }
        Assert.Equal(new[] {
                "I have {currency} and {currency} and another {currency}",
                "I have {currency} and {currency} and another {date}",
                "I have {currency} and {date} and another {currency}",
                "I have {currency} and {date} and another {date}",
                "I have {date} and {currency} and another {currency}",
                "I have {date} and {currency} and another {date}",
                "I have {date} and {date} and another {currency}",
                "I have {date} and {date} and another {date}"
        }, expressions);
    }

    [Fact]
    public void exposes_transforms_in_generated_expression() {
        GeneratedExpression generatedExpression = generator.generateExpressions("I have 2 cukes and 1.5 euro")[0];
        Assert.Equal(typeof(int), generatedExpression.getParameterTypes()[0].getType());
        Assert.Equal(typeof(double), generatedExpression.getParameterTypes()[1].getType());
    }

    [Fact]
    public void matches_parameter_types_with_optional_capture_groups() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "optional-flight",
            "(1st flight)?"));
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "optional-hotel",
            "(1 hotel)?"));
        List<GeneratedExpression> generatedExpressions = generator.generateExpressions("I reach Stage 4: 1st flight -1 hotel");
        Assert.Equal("I reach Stage {int}: {optional-flight} {int} hotel", generatedExpressions[0].getSource());
    }

    [Fact]
    public void generates_at_most_256_expressions() {
        for (int i = 0; i < 4; i++) {
            parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
                "my-type-" + i,
                "[a-z]"));
        }
        // This would otherwise generate 4^11=419430 expressions and consume just shy of 1.5GB.
        Assert.Equal(256, generator.generateExpressions("a b c d e f g h i j k").Count);
    }

    [Fact]
    public void prefers_expression_with_longest_non_empty_match() {
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "exactly-one",
            "[a-z]"));
        parameterTypeRegistry.defineParameterType(new StubParameterType<string>(
            "zero-or-more",
            "[a-z]*"));

        List<GeneratedExpression> generatedExpressions = generator.generateExpressions("a simple step");
        Assert.Equal(2, generatedExpressions.Count);
        Assert.Equal("{exactly-one} {zero-or-more} {zero-or-more}", generatedExpressions[0].getSource());
        Assert.Equal("{zero-or-more} {zero-or-more} {zero-or-more}", generatedExpressions[1].getSource());
    }

    private void assertExpression(String expectedExpression, string[] expectedArgumentNames, String text) {
        GeneratedExpression generatedExpression = generator.generateExpressions(text)[0];
        Assert.Equal(expectedExpression, generatedExpression.getSource());
        Assert.Equal(expectedArgumentNames, generatedExpression.getParameterNames());

        // Check that the generated expression matches the text
        var cucumberExpression = new CucumberExpression(generatedExpression.getSource(), parameterTypeRegistry);
        var match = MatchExpression(cucumberExpression, text);
        match.Should().NotBeNull($"Expected text '{text}' to match generated expression '{generatedExpression.getSource()}'");
        Assert.Equal(expectedArgumentNames.Length, match.Length);
    }

}

File: elixir/test/cucumber/cucumber_expressions/cucumber_expression_generator_test.exs
defmodule Cucumber.CucumberExpressions.CucumberExpressionGeneratorTest do
  use ExUnit.Case, async: true

  alias Cucumber.CucumberExpressions.{
    CucumberExpression,
    GeneratedExpression,
    CucumberExpressionGenerator,
    ParameterType,
    ParameterTypeRegistry
  }

  defp assert_expression(registry, expected_expression, expected_argument_names, text) do
    generated = hd(CucumberExpressionGenerator.generate_expressions(registry, text))
    assert GeneratedExpression.parameter_names(generated) == expected_argument_names
    assert generated.source == expected_expression

    expression = CucumberExpression.compile!(generated.source, registry)
    match = CucumberExpression.match(expression, text)
    refute match == nil, "Expected text '#{text}' to match generated expression"
    assert length(match) == length(expected_argument_names)
  end

  defp define_parameter_type!(registry, options),
    do: ParameterTypeRegistry.define_parameter_type!(registry, ParameterType.new!(options))

  setup do
    {:ok, registry: ParameterTypeRegistry.new()}
  end

  test "documents expression generation", %{registry: registry} do
    undefined_step_text = "I have 2 cucumbers and 1.5 tomato"

    generated =
      hd(CucumberExpressionGenerator.generate_expressions(registry, undefined_step_text))

    assert generated.source == "I have {int} cucumbers and {float} tomato"
    assert Enum.at(generated.parameter_types, 1).type == :float
  end

  test "generates expression for no args", %{registry: registry} do
    assert_expression(registry, "hello", [], "hello")
  end

  test "generates expression with escaped left parenthesis", %{registry: registry} do
    assert_expression(registry, "\\(iii)", [], "(iii)")
  end

  test "generates expression with escaped left curly brace", %{registry: registry} do
    assert_expression(registry, "\\{iii}", [], "{iii}")
  end

  test "generates expression with escaped slashes", %{registry: registry} do
    assert_expression(
      registry,
      "The {int}\\/{int}\\/{int} hey",
      ["int", "int2", "int3"],
      "The 1814/05/17 hey"
    )
  end

  test "generates expression for int float arg", %{registry: registry} do
    assert_expression(
      registry,
      "I have {int} cukes and {float} euro",
      ["int", "float"],
      "I have 2 cukes and 1.5 euro"
    )
  end

  test "generates expression for strings", %{registry: registry} do
    assert_expression(
      registry,
      "I like {string} and {string}",
      ["string", "string2"],
      "I like \"bangers\" and 'mash'"
    )
  end

  test "generates expression with % sign", %{registry: registry} do
    assert_expression(registry, "I am {int}% foobar", ["int"], "I am 20% foobar")
  end

  test "generates expression for just int", %{registry: registry} do
    assert_expression(registry, "{int}", ["int"], "99999")
  end

  test "numbers only second argument when builtin type is not reserved keyword", %{
    registry: registry
  } do
    assert_expression(
      registry,
      "I have {int} cukes and {int} euro",
      ["int", "int2"],
      "I have 2 cukes and 5 euro"
    )
  end

  test "numbers only second argument when type is not reserved keyword", %{registry: registry} do
    registry =
      define_parameter_type!(registry,
        name: "currency",
        regexps: "[A-Z]{3}",
        type: :currency,
        prefer_for_regexp_match: true
      )

    assert_expression(
      registry,
      "I have a {currency} account and a {currency} account",
      ["currency", "currency2"],
      "I have a EUR account and a GBP account"
    )
  end

  test "exposes parameters in a generated expression", %{registry: registry} do
    generated =
      hd(
        CucumberExpressionGenerator.generate_expressions(registry, "I have 2 cukes and 1.5 euro")
      )

    assert Enum.map(generated.parameter_types, & &1.type) == [:integer, :float]
  end

  test "matches parameter types with optional capture groups", %{registry: registry} do
    registry =
      registry
      |> define_parameter_type!(name: "optional-flight", regexps: ~r/(1st flight)?/)
      |> define_parameter_type!(name: "optional-hotel", regexps: ~r/(1 hotel)?/)

    generated =
      hd(
        CucumberExpressionGenerator.generate_expressions(
          registry,
          "I reach Stage 4: 1st flight -1 hotel"
        )
      )

    # While you would expect this to be `I reach Stage {int}: {optional-flight} -{optional-hotel}`
    # the `-1` causes {int} to match just before {optional-hotel}.
    assert generated.source == "I reach Stage {int}: {optional-flight} {int} hotel"
  end

  test "generates at most 256 expressions", %{registry: registry} do
    registry =
      Enum.reduce(0..3, registry, fn i, registry ->
        define_parameter_type!(registry, name: "my-type-#{i}", regexps: ~r/([a-z] )*?[a-z]/)
      end)

    # This would otherwise generate 4^11=4194304 expressions.
    expressions =
      CucumberExpressionGenerator.generate_expressions(registry, "a s i m p l e s t e p")

    assert length(expressions) == 256
  end

  test "prefers expression with longest non empty match", %{registry: registry} do
    registry =
      registry
      |> define_parameter_type!(name: "zero-or-more", regexps: ~r/[a-z]*/)
      |> define_parameter_type!(name: "exactly-one", regexps: ~r/[a-z]/)

    expressions = CucumberExpressionGenerator.generate_expressions(registry, "a simple step")
    assert length(expressions) == 2
    assert Enum.at(expressions, 0).source == "{exactly-one} {zero-or-more} {zero-or-more}"
    assert Enum.at(expressions, 1).source == "{zero-or-more} {zero-or-more} {zero-or-more}"
  end

  describe "word boundaries" do
    setup %{registry: registry} do
      {:ok, registry: define_parameter_type!(registry, name: "direction", regexps: ~r/(up|down)/)}
    end

    test "does not suggest a parameter at the beginning of a word", %{registry: registry} do
      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "When I download a picture"
               )
             ).source ==
               "When I download a picture"
    end

    test "does not suggest a parameter inside a word", %{registry: registry} do
      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "When I watch the muppet show"
               )
             ).source ==
               "When I watch the muppet show"
    end

    test "does not suggest a parameter at the end of a word", %{registry: registry} do
      assert hd(
               CucumberExpressionGenerator.generate_expressions(registry, "When I create a group")
             ).source ==
               "When I create a group"
    end

    test "suggests a parameter for a full word", %{registry: registry} do
      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "When I go down the road"
               )
             ).source ==
               "When I go {direction} the road"

      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "When I walk up the hill"
               )
             ).source ==
               "When I walk {direction} the hill"

      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "up the hill, the road goes down"
               )
             ).source ==
               "{direction} the hill, the road goes {direction}"
    end

    test "suggests a parameter wrapped around punctuation characters", %{registry: registry} do
      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "When direction is:down"
               )
             ).source ==
               "When direction is:{direction}"

      assert hd(
               CucumberExpressionGenerator.generate_expressions(
                 registry,
                 "Then direction is down."
               )
             ).source ==
               "Then direction is {direction}."
    end
  end
end
